### PR TITLE
[AMD] Add MoE weights and scales padding

### PR DESCRIPTION
**Pull request by mqhc2020**
## Motivation

Right now, Aiter MoE requires weights and scales to align with a fixed number. Since some models have intermediate sizes that don't fit this rule, we need to add extra padding to the weights so they can be processed by the Fused MoE.

## Modifications

Add padding for the weights. Below listed are the models and configurations that has been verified with:
1. Qwen/Qwen3-235B-A22B-Instruct-2507-FP8 : TP=8
2. amd/Qwen3-235B-A22B-Instruct-2507-MXFP4: TP=8
3. zai-org/GLM-4.7: TP=8

## Accuracy Tests

We run GSM8K accuracy test for models below:

Qwen/Qwen3-235B-A22B-Instruct-2507-FP8 (TP8):  

Accuracy: 0.980
Invalid: 0.000
Latency: 10.525 s
Output throughput: 2663.037 token/s

amd/Qwen3-235B-A22B-Instruct-2507-MXFP4 (TP8):

Accuracy: 0.940
Invalid: 0.000
Latency: 23.853 s
Output throughput: 7488.441 token/s
zai-org/GLM-4.7 (TP8):

(better than VLLM TP4 on model card)

Accuracy: 0.975
Invalid: 0.000
Latency: 7.349 s
Output throughput: 2683.189 token/s

## Benchmarking and Profiling

Baseline benchmark for amd/Qwen3-235B-A22B-Instruct-2507-mxfp4 (TP4) and Qwen/Qwen3-235B-A22B-Instruct-2507-FP8 (TP8):
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
5. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
6. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
7. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/model_executor/model_runner.py [hunk 1]
```diff
@@ -732,7 +721,9 @@ def check_quantized_moe_compatibility(self):
                     f"moe_intermediate_size {moe_intermediate_size} must be divisible by moe_tp_size ({moe_tp_size}) which is tp_size ({self.tp_size}) divided by moe_ep_size ({self.moe_ep_size})."
                 )
 
-            if (moe_intermediate_size // moe_tp_size) % weight_block_size_n != 0:
+            if (
+                moe_intermediate_size // moe_tp_size
+            ) % weight_block_size_n != 0 and not _is_hip:
```
Comment (sogalin) [inline python/sglang/srt/model_executor/model_runner.py]: Hello @mqhc2020,
With this modification, it looks like we are going to ignore all checks in SGL when _is_hip is True?

File: python/sglang/srt/layers/quantization/fp8.py [hunk 2]
```diff
@@ -739,27 +716,42 @@ def create_weights(
...
+                # Required by column parallel or enabling merged weights
+                if intermediate_size_per_partition % block_n != 0:
                     raise ValueError(
-                        f"The input_size of down's weight = "
+                        f"The output_size of gate's and up's weight = "
                         f"{intermediate_size_per_partition} is not divisible by "
-                        f"weight quantization block_k = {block_k}."
+                        f"weight quantization block_n = {block_n}."
                     )
+                if tp_size > 1:
```
Comment (yichiche) [inline python/sglang/srt/layers/quantization/fp8.py:810]: Do aiter support TP > 1 case if using block_quant?

Comment (mqhc2020) [reply python/sglang/srt/layers/quantization/fp8.py:810]: Yes, I think the sanity check should be kept for aiter MoE too.

**Review by yichiche — APPROVED** [review]
LGTM

**Comment by mqhc2020** [comment]
I found a problem myself, running Grok 1. Trying to fix that.

**Review by sogalin — APPROVED** [review]

**Review by kkHuang-amd — COMMENTED** [review]
Could we add a utile function to do padding and remove the padding_size declaration in the "python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py", the place is weird.

**Comment by mqhc2020** [comment]
> @mqhc2020 , Is accuracy expected 0.835 with amd/Qwen3-235B-A22B-Instruct-2507-mxfp4 (TP4)?
> 
> Accuracy: 0.835 Invalid: 0.000 Latency: 13.401 s Output throughput: 2009.483 token/s

@sogalin  this patch involves various models. I can't guarantee the accuracy of the models can be totally fine after the 128-aligned problem is fixed by this PR. That might be another task later.

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py [hunk 3]
```diff
@@ -99,7 +110,7 @@ def create_weights(
             torch.empty(
                 num_experts,
                 hidden_size,
-                intermediate_size_per_partition // 2,
+                w2_down_dim,
```
Comment (fxmarty-amd) [inline python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py:113]: There is no issue with padding here? Won't we load data into wrong columns?

Comment (fxmarty-amd) [inline python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py:113]: Also: why is `w2_weight` padded but not `w2_weight_scale`?

File: python/sglang/srt/layers/moe/utils.py [hunk 4]
```diff
@@ -341,3 +336,30 @@ class RoutingMethodType(IntEnum):
...
+
+    if is_aiter_moe:
+        padding_size = get_moe_padding_size(True)
+        align_aiter = lambda n: ((n + padding_size - 1) // padding_size) * padding_size
+        is_padded = (w2_down_dim % padding_size) > 0
+        if is_padded:
+            w2_down_dim = align_aiter(w2_down_dim)
+        # packed
+        if hasattr(torch, "float4_e2m1fn_x2") and is_packed:
+            w13_up_dim *= 2
```
Comment (fxmarty-amd) [inline python/sglang/srt/layers/moe/utils.py]: Why not `w2_down_dim` as well here? Is this for fp4 packing on one byte or for `gate_proj`/`up_proj` packing only?

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py (hunk 3 above)
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py:113]: Ah no this is handled in `python/sglang/srt/layers/moe/fused_moe_triton/layer.py`.

Personally, I would prefer to have padding handled in `process_weights_after_loading` (e.g. it makes it easier for online quantization), but both are fine though.

Comment (mqhc2020) [reply python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py:113]: Yes, loading happens in layer.py. It basically copy tensors with 128-aligned inter_dim and some GPUs have "dummy" tensors, which won't contribute to the logits.

File: python/sglang/srt/layers/moe/utils.py (hunk 4 above)
Comment (mqhc2020) [reply python/sglang/srt/layers/moe/utils.py]: The alignment unit is "byte" and two fp4s are packed into one byte.

File: python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py (hunk 3 above)
Comment (mqhc2020) [reply python/sglang/srt/layers/quantization/quark/schemes/quark_w4a4_mxfp4_moe.py:113]: Yes you are right and we are on it. This should have been fixed in latest commit.

**Review by fxmarty-amd — APPROVED** [review]
LGTM.

File: python/sglang/srt/layers/moe/utils.py [hunk 5]
```diff
@@ -341,3 +344,30 @@ class RoutingMethodType(IntEnum):
     TopK = (5,)
     # Unspecified
     Unspecified = 6
+
+
+def get_moe_padding_size(is_aiter_moe):
+    if is_aiter_moe:
+        return 128
```
Comment (BowenBao) [inline python/sglang/srt/layers/moe/utils.py:351]: Hi @mqhc2020 , I recently looked at something similar on vllm side https://github.com/vllm-project/vllm/pull/34285, it was mentioned that [256 pad size might have better performance](https://github.com/vllm-project/vllm/pull/34285#discussion_r2855710524). Could you check if this is the case too?

**Comment by HaiShaw** [comment]
@saienduri please help to look at time out issue from new runner? cc @bingxche @yctseng0211 
@BowenBao will merge now. @mqhc2020 please try out padding 256 and file another PR if that is better.
